identity/src/main/java/com/android/identity/credential/NameSpacedData.java

```diff
@@ -0,0 +1,390 @@
+            LinkedHashMap<String, byte[]> namespace = mMap.get(namespaceName);
+            for (String dataElementName : namespace.keySet()) {
+                byte[] dataElementValue = namespace.get(dataElementName);
+                ByteString taggedBstr = new ByteString(dataElementValue);
+                taggedBstr.setTag(new Tag(24));
+                innerMapBuilder.put(new UnicodeString(dataElementName), taggedBstr);
+            }
+        }
+        return builder.build().get(0);
+    }
+
+    /**
+     * Encodes the given {@link NameSpacedData} as CBOR.
+     *
+     * <p>The encoding uses the following CDDL:
+     * <pre>
+     *     NameSpacedCbor = {
+     *         + NameSpaceName =&gt; DataElements
+     *     }
+     *
+     *     NameSpaceName = tstr
+     *
+     *     DataElements = [
+     *         + DataElementName =&gt; DataElementValueBytes
+     *     ]
+     *
+     *     DataElementName = tstr
+     *     DataElementValue = any
+     *     DataElementValueBytes = #6.24(bstr .cbor DataElementValue)
+     * </pre>
+     *
+     * <p>Here's an example of CBOR conforming to the above CDDL printed in diagnostic form:
+     * <pre>
+     *     {
+     *         "org.iso.18013.5.1" : {
+     *             "given_name" : 24(&lt;&lt; "Erika" &gt;&gt;),
+     *             "family_name" : 24(&lt;&lt; "Mustermann" &gt;&gt;),
+     *         },
+     *         "org.iso.18013.5.1.aamva" : {
+     *             "organ_donor" : 24(&lt;&lt; 1 &gt;&gt;)
+     *         }
+     *     }
+     * </pre>
+     *
+     * <p>Name spaces and data elements will be in the order they were inserted using
+     * at construction time, either using the {@link Builder} or
+     * {@link #fromEncodedCbor(byte[])}).
+     *
+     * @return the bytes of the encoding describe above.
+     */
+    public @NonNull byte[] encodeAsCbor() {
+        return Util.cborEncode(toCbor());
+    }
+
+    /**
+     * A builder for {@link NameSpacedData}.
+     */
+    public static class Builder {
+        LinkedHashMap<String, LinkedHashMap<String, byte[]>> mMap = new LinkedHashMap<>();
+
+        /**
+         * Constructor.
+         */
+        public Builder() {}
+
+        /**
+         * Adds a raw CBOR value to the builder.
+         *
+         * @param nameSpaceName the name space name.
+         * @param dataElementName the data element name.
+         * @param value the bytes of the CBOR.
+         * @return the builder.
+         */
+        public @NonNull Builder putEntry(@NonNull String nameSpaceName,
+                                         @NonNull String dataElementName,
+                                         @NonNull byte[] value) {
+            LinkedHashMap<String, byte[]> innerMap = mMap.get(nameSpaceName);
+            if (innerMap == null) {
+                innerMap = new LinkedHashMap<>();
+                mMap.put(nameSpaceName, innerMap);
+            }
+            // TODO: validate/verify that value is proper CBOR.
```

Comment:
could this be done as part of this PR?

Reply:
I was putting it off because I'm on the fence about it. The thing is that it's pretty expensive to check for this... I think we should just note in the docs that we don't check for it and if it's untrusted data the caller probably should. I'll do that.